File: spring/src/main/java/cardsproject/domain/marketplace/TradeBid.java
package cardsproject.domain.marketplace;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "trade_bids")
public class TradeBid {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private BigDecimal amount = BigDecimal.ZERO;
    private LocalDateTime placedAt;
    private Boolean isWinning = false;

    @Column(name = "listing_id")
    private Long listingId;
    @Column(name = "bidder_id")
    private Long bidderId;

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public BigDecimal getAmount() { return amount; }
    public void setAmount(BigDecimal amount) { this.amount = amount; }
    public LocalDateTime getPlacedAt() { return placedAt; }
    public void setPlacedAt(LocalDateTime placedAt) { this.placedAt = placedAt; }
    public Boolean getIsWinning() { return isWinning; }
    public void setIsWinning(Boolean isWinning) { this.isWinning = isWinning; }
    public Long getListingId() { return listingId; }
    public void setListingId(Long listingId) { this.listingId = listingId; }
    public Long getBidderId() { return bidderId; }
    public void setBidderId(Long bidderId) { this.bidderId = bidderId; }

    // ── Business operations ──────────────────────────────────────────
    public Boolean outbidBy(java.math.BigDecimal newAmount) {
        throw new UnsupportedOperationException("outbidBy not implemented");
    }
}

File: symfony/src/Entity/Marketplace/TradeBid.php
<?php

namespace App\Entity\Marketplace;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Annotation\Groups;
use App\Repository\Marketplace\TradeBidRepository;
use App\Entity\Players\Player;

#[ORM\Entity(repositoryClass: TradeBidRepository::class)]
#[ORM\Table(name: 'trade_bid')]
class TradeBid
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    #[Groups(['tradeBid:read'])]
    private ?int $id = null;

    #[ORM\Column(type: 'decimal', precision: 10, scale: 2)]
    #[Groups(['tradeBid:read', 'tradeBid:write'])]
    private string $amount = '0.00';

    #[ORM\Column(type: 'datetime', nullable: true)]
    #[Groups(['tradeBid:read', 'tradeBid:write'])]
    private ?\DateTimeInterface $placedAt = null;

    #[ORM\Column(type: 'boolean')]
    #[Groups(['tradeBid:read', 'tradeBid:write'])]
    private bool $isWinning = false;

    #[ORM\ManyToOne(targetEntity: TradeListing::class, inversedBy: 'bids')]
    #[ORM\JoinColumn(nullable: false)]
    private ?TradeListing $listing = null;

    #[ORM\ManyToOne(targetEntity: Player::class, inversedBy: 'bids')]
    #[ORM\JoinColumn(nullable: false)]
    private ?Player $bidder = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getAmount(): string
    {
        return $this->amount;
    }

    public function setAmount(string $amount): static
    {
        $this->amount = $amount;
        return $this;
    }

    public function getPlacedAt(): ?\DateTimeInterface
    {
        return $this->placedAt;
    }

    public function setPlacedAt(?\DateTimeInterface $placedAt): static
    {
        $this->placedAt = $placedAt;
        return $this;
    }

    public function getIsWinning(): bool
    {
        return $this->isWinning;
    }

    public function setIsWinning(bool $isWinning): static
    {
        $this->isWinning = $isWinning;
        return $this;
    }

    #[Groups(['tradeBid:read'])]
    public function getListingId(): ?int
    {
        return $this->listing?->getId();
    }

    public function getListing(): ?TradeListing
    {
        return $this->listing;
    }

    public function setListing(?TradeListing $listing): static
    {
        $this->listing = $listing;
        return $this;
    }

    #[Groups(['tradeBid:read'])]
    public function getBidderId(): ?int
    {
        return $this->bidder?->getId();
    }

    public function getBidder(): ?Player
    {
        return $this->bidder;
    }

    public function setBidder(?Player $bidder): static
    {
        $this->bidder = $bidder;
        return $this;
    }

    // ── Validation rules ─────────────────────────────────────────────
    #[\Symfony\Component\Validator\Constraints\IsTrue(message: "Bid amount must be greater than zero")]
    public function isAmountPositiveValid(): bool
    {
        return ($this->getAmount() === null || (float)$this->getAmount() > (float)0);
    }

    // ── Business operations ──────────────────────────────────────────

    public function outbidBy($newAmount): void
    {
        throw new \RuntimeException('outbid_by not implemented');
    }

    public function retract(): void
    {
        throw new \RuntimeException('retract not implemented');
    }

}
